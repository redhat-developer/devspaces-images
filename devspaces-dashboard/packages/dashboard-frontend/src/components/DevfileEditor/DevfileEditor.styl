.devfile-editor
  width 100%
  height 100%
  min-height 50vh
  .monaco
    position relative
    width 100%
    height calc(100% - 38px)
    min-height inherit
    .monaco-editor
      border 1px solid #cecece
      position absolute
      top 0
      left 0
      right 0
      bottom 0
    .devfile-editor-decoration
      background-color alpha(#f1c050, 0.5)
  .error
    color #e22812
    font-size 15px
    height 16px
  a
    color #3782c4
    font-size 15px
    line-height 25px
    text-decoration none
