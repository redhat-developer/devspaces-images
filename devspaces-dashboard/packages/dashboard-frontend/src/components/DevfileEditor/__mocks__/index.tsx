/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React, { ChangeEvent, ClipboardEvent } from 'react';
import { safeDump } from 'js-yaml';
import devfileApi from '../../../services/devfileApi';
import stringify from '../../../services/helpers/editor';

type Props = {
  devfile: che.WorkspaceDevfile | devfileApi.Devfile;
  decorationPattern?: string;
  onChange: (newValue: string, isValid: boolean) => void;
  isReadonly?: boolean;
};
type State = {
  errorMessage: string;
};

export default class DevfileEditor extends React.PureComponent<Props, State> {
  private editor: HTMLTextAreaElement;

  constructor(props: Props) {
    super(props);

    this.state = {
      errorMessage: '',
    };
  }

  public updateContent(devfile: che.WorkspaceDevfile | devfileApi.Devfile): void {
    if (!this.editor) {
      return;
    }
    this.editor.value = stringify(devfile);
  }

  public componentDidMount(): void {
    const element = document.querySelector('#devfile-editor');
    if (element) {
      const value = stringify(this.props.devfile);
      this.editor = element as HTMLTextAreaElement;
      this.editor.value = value;
    }
  }

  public render(): React.ReactElement {
    const { devfile } = this.props;
    const content = safeDump(devfile);

    const onChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
      const newContent = event.currentTarget.value;
      this.onChange(newContent, true);
    };

    const onPaste = (event: ClipboardEvent<HTMLTextAreaElement>) => {
      const newContent = event.clipboardData?.getData('Text') || '';
      this.onChange(newContent, true);
    };

    return (
      <div className="devfile-editor">
        <div className="monaco">
          <textarea
            id="devfile-editor"
            onChange={onChange}
            onPaste={onPaste}
            value={content}
          ></textarea>
        </div>
        <div className="error"></div>
        <a target="_blank" rel="noopener noreferrer">
          Devfile Documentation
        </a>
      </div>
    );
  }

  private onChange(newValue: string, isValid: boolean): void {
    this.props.onChange(newValue, isValid);
  }
}
