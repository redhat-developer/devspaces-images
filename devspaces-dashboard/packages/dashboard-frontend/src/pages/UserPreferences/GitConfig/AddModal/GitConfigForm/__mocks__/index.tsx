/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/AddModal/GitConfigForm';
import * as GitConfigStore from '@/store/GitConfig';

export class GitConfigForm extends React.PureComponent<Props> {
  render() {
    const { onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { name: 'User One' } } as GitConfigStore.GitConfig, false)
          }
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { email: 'user-1@chetest.com', name: 'User One' } }, true)
          }
        />
      </div>
    );
  }
}
