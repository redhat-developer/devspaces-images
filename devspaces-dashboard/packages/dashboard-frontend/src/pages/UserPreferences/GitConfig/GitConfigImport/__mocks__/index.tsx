/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/GitConfigImport';

export class GitConfigImport extends React.PureComponent<Props> {
  render() {
    const { content, onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          value={content}
          onClick={() => onChange('[user]\n\tname = User One\n', false)}
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange('[user]\n\tname = User One\n\temail = user-1@chetest.com\n', true)
          }
        />
      </div>
    );
  }
}
