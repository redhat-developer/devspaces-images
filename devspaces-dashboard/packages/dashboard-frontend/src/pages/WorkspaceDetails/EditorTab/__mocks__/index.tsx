/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import common from '@eclipse-che/common';
import React from 'react';
import { Props, State } from '..';
import devfileApi from '../../../../services/devfileApi';
import { Workspace } from '../../../../services/workspace-adapter';

export default class EditorTab extends React.PureComponent<Props, State> {
  cancelChanges: () => void;

  constructor(props: Props) {
    super(props);

    this.state = {
      devfile: {} as devfileApi.Devfile,
      hasChanges: false,
      isDevfileValid: true,
      hasRequestErrors: false,
      currentRequestError: '',
      isExpanded: false,
      showDevfileV2ConfirmationModal: false,
    };

    this.cancelChanges = () => {
      // no-op
    };
  }

  async handleOnSave(workspace: Workspace) {
    try {
      await this.props.onSave(workspace);
    } catch (e) {
      this.setState({
        currentRequestError: common.helpers.errors.getMessage(e),
      });
    }
  }

  render() {
    return (
      <div>
        Fake Editor Tab
        <button onClick={() => this.handleOnSave(this.props.workspace)}>Save</button>
        <button onClick={() => this.props.onDevWorkspaceWarning()}>Warning</button>
        <span data-testid="current-request-error">{this.state.currentRequestError}</span>
      </div>
    );
  }
}
