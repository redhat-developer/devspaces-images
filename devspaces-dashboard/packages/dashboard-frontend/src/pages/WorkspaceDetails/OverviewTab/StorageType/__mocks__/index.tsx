/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Props, State } from '..';

export default class StorageTypeFormGroup extends React.PureComponent<Props, State> {
  render() {
    return <div>Fake Storage Type Form</div>;
  }
}
