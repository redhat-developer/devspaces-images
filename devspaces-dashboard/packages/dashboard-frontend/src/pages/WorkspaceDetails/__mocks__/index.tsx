/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Props, State } from '..';

export const mockShowAlert = jest.fn();
export class WorkspaceDetails extends React.PureComponent<Props, State> {
  render() {
    if (!this.props.workspace) {
      return <div>Not found</div>;
    }
    const workspace = this.props.workspace;
    return (
      <div className="WorkspaceDetails">
        Workspace Details Page
        <div data-testid="props-is-loading">{this.props.isLoading}</div>
        <div data-testid="props-old-workspace-path">
          {this.props.oldWorkspaceLocation?.pathname}
        </div>
        <div data-testid="props-show-convert-button">
          {this.props.showConvertButton === true ? 'true' : 'false'}
        </div>
        <div data-testid="props-workspace-id">{this.props.workspace.id}</div>
        <div data-testid="props-workspace-name">{this.props.workspace.name}</div>
        <div data-testid="props-workspaces-link">{this.props.workspacesLink}</div>
        <button onClick={() => this.props.onSave(workspace)}>Save</button>
        <button onClick={() => this.props.onConvert(workspace)}>Convert</button>
      </div>
    );
  }
}
