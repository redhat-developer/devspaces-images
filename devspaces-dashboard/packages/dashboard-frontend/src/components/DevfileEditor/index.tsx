/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { AppState } from '../../store';
import { DisposableCollection } from '../../services/helpers/disposable';
import {
  ProtocolToMonacoConverter,
  MonacoToProtocolConverter,
} from 'monaco-languageclient/lib/monaco-converter';
import { languages, editor, Range, Position, IRange } from 'monaco-editor-core';
import {
  TextDocument,
  getLanguageService,
  LanguageService,
  CompletionItem,
} from 'yaml-language-server';
import { initDefaultEditorTheme } from '../../services/monacoThemeRegister';
import stringify, { language, conf } from '../../services/helpers/editor';
import $ from 'jquery';
import { merge, isMatch } from 'lodash';
import devfileApi from '../../services/devfileApi';
import { selectDevfileSchema } from '../../store/DevfileRegistries/selectors';
import { selectPlugins } from '../../store/Plugins/chePlugins/selectors';
import { selectBranding } from '../../store/Branding/selectors';

import './DevfileEditor.styl';

const LANGUAGE_ID = 'yaml';
const YAML_SERVICE = 'yamlService';
const MONACO_CONFIG: editor.IStandaloneEditorConstructionOptions = {
  language: 'yaml',
  wordWrap: 'on',
  lineNumbers: 'on',
  scrollBeyondLastLine: false,
  readOnly: false,
};

(self as any).MonacoEnvironment = {
  getWorkerUrl: () => './editor.worker.js',
};

type Props = MappedProps & {
  devfile: che.WorkspaceDevfile | devfileApi.Devfile;
  decorationPattern?: string;
  onChange: (newValue: string, isValid: boolean) => void;
  isReadonly?: boolean;
  additionSchema?: { [key: string]: any };
};
type State = {
  errorMessage: string;
};

export class DevfileEditor extends React.PureComponent<Props, State> {
  public static EDITOR_THEME: string | undefined;
  private readonly toDispose = new DisposableCollection();
  private handleResize: () => void;
  private editor: editor.IStandaloneCodeEditor;
  private readonly yamlService: LanguageService;
  private m2p = new MonacoToProtocolConverter();
  private p2m = new ProtocolToMonacoConverter();
  private createDocument = (model): TextDocument =>
    TextDocument.create(
      'inmemory://model.yaml',
      model.getModeId(),
      model.getVersionId(),
      model.getValue(),
    );
  private skipNextOnChange: boolean;

  constructor(props: Props) {
    super(props);

    this.state = {
      errorMessage: '',
    };

    // lazy initialization
    if (!window[YAML_SERVICE]) {
      this.yamlService = getLanguageService(() => Promise.resolve(''), {
        resolveRelativePath: () => '',
      });
      window[YAML_SERVICE] = this.yamlService;
    } else {
      this.yamlService = window[YAML_SERVICE];
      this.updateSchema();
      return;
    }
    if (!DevfileEditor.EDITOR_THEME) {
      // define the default
      DevfileEditor.EDITOR_THEME = initDefaultEditorTheme();
    }
    if (
      languages?.getLanguages &&
      languages.getLanguages().find(language => language.id === LANGUAGE_ID) === undefined
    ) {
      // register the YAML language with Monaco
      languages.register({
        id: LANGUAGE_ID,
        extensions: ['.yaml', '.yml'],
        aliases: ['YAML', 'yaml'],
        mimetypes: ['application/yaml'],
      });
      languages.setMonarchTokensProvider(LANGUAGE_ID, language);
      languages.setLanguageConfiguration(LANGUAGE_ID, conf);
    }
    // register language server providers
    this.registerLanguageServerProviders();
    this.updateSchema();
  }

  private updateSchema(): void {
    let jsonSchema = this.props.devfileSchema ? Object.assign({}, this.props.devfileSchema) : {};
    const additionSchema = this.props.additionSchema;
    if (additionSchema) {
      if (jsonSchema?.oneOf) {
        jsonSchema.oneOf = jsonSchema.oneOf.map(schema => merge({}, schema, additionSchema));
      } else {
        jsonSchema = merge({}, jsonSchema, additionSchema);
      }
    }
    const schemas = [{ uri: 'inmemory:yaml', fileMatch: ['*'], schema: jsonSchema }];
    this.yamlService.configure({ validate: true, hover: true, schemas, completion: true });
  }

  public updateContent(devfile: che.WorkspaceDevfile | devfileApi.Devfile): void {
    if (!this.editor) {
      return;
    }
    this.skipNextOnChange = true;
    const doc = this.editor.getModel();
    doc?.setValue(stringify(devfile));
  }

  public componentDidUpdate(prevProps: Props): void {
    if (this.handleResize) {
      this.handleResize();
    }
    if (!isMatch(prevProps.additionSchema || {}, this.props.additionSchema || {})) {
      this.updateSchema();
    }
  }

  // This method is called when the component is first added to the document
  public componentDidMount(): void {
    const element = $('.devfile-editor .monaco').get(0);
    if (element) {
      const value = stringify(this.props.devfile);
      this.editor = editor.create(
        element,
        Object.assign(
          { value },
          {
            ...MONACO_CONFIG,
            readOnly: !!this.props.isReadonly,
          },
        ),
      );
      const doc = this.editor.getModel();
      doc?.updateOptions({ tabSize: 2 });

      const handleResize = (): void => {
        const layout = { height: element.offsetHeight, width: element.offsetWidth };
        // if the element is hidden
        if (layout.height === 0 || layout.width === 0) {
          return;
        }
        this.editor.layout(layout);
      };
      this.handleResize = handleResize;
      window.addEventListener('resize', handleResize);
      this.toDispose.push({
        dispose: () => {
          window.removeEventListener('resize', handleResize);
        },
      });

      let oldDecorationIds: string[] = []; // Array containing previous decorations identifiers.
      const updateDecorations = (): void => {
        if (this.props.decorationPattern) {
          oldDecorationIds = this.editor.deltaDecorations(oldDecorationIds, this.getDecorations());
        }
      };
      updateDecorations();
      doc?.onDidChangeContent(() => {
        updateDecorations();
        this.onChange(this.editor.getValue(), true);
      });
      // init language server validation
      this.initLanguageServerValidation(this.editor);
    }

    const handleMessage = (event: MessageEvent): void => {
      if (typeof event.data !== 'string') {
        return;
      }
      const { data } = event;
      if (
        (data === 'show-navbar' || data === 'hide-navbar' || data === 'toggle-navbar') &&
        this.handleResize
      ) {
        this.handleResize();
      }
    };
    window.addEventListener('message', handleMessage, false);
    this.toDispose.push({
      dispose: () => {
        window.removeEventListener('message', handleMessage, false);
      },
    });
  }

  // This method is called when the component is removed from the document
  public componentWillUnmount(): void {
    this.toDispose.dispose();
  }

  public render(): React.ReactElement {
    const href = this.props.branding.docs.devfile;
    const { errorMessage } = this.state;

    return (
      <div className="devfile-editor">
        <div className="monaco">&nbsp;</div>
        <div className="error">{errorMessage}</div>
        <a target="_blank" rel="noopener noreferrer" href={href}>
          Devfile Documentation
        </a>
      </div>
    );
  }

  private getDecorations(): editor.IModelDecoration[] {
    const decorations: editor.IModelDecoration[] = [];
    if (this.props.decorationPattern) {
      const decorationRegExp = new RegExp(this.props.decorationPattern, 'img');
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const model = this.editor.getModel()!;
      const value = this.editor.getValue();
      let match = decorationRegExp.exec(value);
      while (match) {
        const startPosition = model.getPositionAt(match.index);
        const endPosition = model.getPositionAt(match.index + match[0].length);
        const range = new Range(
          startPosition.lineNumber,
          startPosition.column,
          endPosition.lineNumber,
          endPosition.column,
        );
        const options: editor.IModelDecorationOptions = {
          inlineClassName: 'devfile-editor-decoration',
        };
        decorations.push({
          range,
          options,
        } as editor.IModelDecoration);
        match = decorationRegExp.exec(value);
      }
    }
    return decorations;
  }

  private onChange(newValue: string, isValid: boolean): void {
    if (this.skipNextOnChange) {
      this.skipNextOnChange = false;
      return;
    }
    this.props.onChange(newValue, isValid);
  }

  private registerLanguageServerProviders(): void {
    const createDocument = this.createDocument;
    const yamlService = this.yamlService;
    const m2p = this.m2p;
    const p2m = this.p2m;

    const sortTextCache = {};
    const createSortText = (pluginId: string): string => {
      if (sortTextCache[pluginId]) {
        return sortTextCache[pluginId];
      }
      const [publisher, name, version] = pluginId.split('/');
      const semverRE = /^\d+?\.\d+?\.\d+?/;
      let versionText: string;
      if (semverRE.test(version)) {
        let [major, minor, rest] = version.split('.');
        major = addLeadingZeros(major);
        minor = addLeadingZeros(minor);
        rest = rest.replace(/^(\d+)/, addLeadingZeros);
        versionText = `${major}.${minor}.${rest}`;
      } else {
        let prefix = '-';
        if (version === 'latest') {
          prefix += '0';
        } else if (version === 'next') {
          prefix += '1';
        } else if (version === 'nightly') {
          prefix += '2';
        }
        versionText = `${prefix}${version}`;
      }
      const sortText = `${publisher}/${name}/${versionText}`;
      sortTextCache[pluginId] = sortText;
      return sortText;
    };
    const addLeadingZeros = (number: string): string => {
      while (number.length < 4) {
        number = '0' + number;
      }
      return number;
    };

    languages.registerCompletionItemProvider(LANGUAGE_ID, {
      provideCompletionItems(model: editor.ITextModel, position: Position) {
        const document = createDocument(model);
        return yamlService
          .doComplete(document, m2p.asPosition(position.lineNumber, position.column), true)
          .then(list => {
            const completionResult = p2m.asCompletionResult(
              list as any,
              {
                startColumn: position.column,
                startLineNumber: position.lineNumber,
                endColumn: position.column,
                endLineNumber: position.lineNumber,
              } as IRange,
            );
            if (!completionResult || !completionResult.suggestions) {
              return completionResult;
            }
            // convert completionResult into suggestions
            const defaultInsertTextRules = languages.CompletionItemInsertTextRule.InsertAsSnippet;
            const suggestions = completionResult.suggestions.map(suggestion => {
              return Object.assign(suggestion, {
                insertTextRules: suggestion.insertTextRules
                  ? suggestion.insertTextRules
                  : defaultInsertTextRules,
                sortText: createSortText(suggestion.insertText),
              });
            });
            return { suggestions };
          });
      },
      async resolveCompletionItem(model, range, item) {
        return (yamlService as any)
          .doResolve(m2p.asCompletionItem(item))
          .then((result: CompletionItem) => p2m.asCompletionItem(result as any, range));
      },
    } as any);
    languages.registerDocumentSymbolProvider(LANGUAGE_ID, {
      provideDocumentSymbols(model: editor.ITextModel) {
        return p2m.asSymbolInformations(yamlService.findDocumentSymbols(createDocument(model)));
      },
    });
    languages.registerHoverProvider(LANGUAGE_ID, {
      async provideHover(model: editor.ITextModel, position: Position) {
        const hover = await yamlService.doHover(
          createDocument(model),
          m2p.asPosition(position.lineNumber, position.column),
        );
        return p2m.asHover(hover);
      },
    });
  }

  private initLanguageServerValidation(_editor: editor.IStandaloneCodeEditor): void {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const model = _editor.getModel()!;
    let validationTimer: number;

    model.onDidChangeContent(() => {
      const document = this.createDocument(model);
      this.setState({ errorMessage: '' });
      if (validationTimer) {
        clearTimeout(validationTimer);
      }
      validationTimer = window.setTimeout(() => {
        this.yamlService.doValidation(document, false).then(diagnostics => {
          const markers = this.p2m.asDiagnostics(diagnostics) as any;
          let errorMessage = '';
          if (Array.isArray(markers) && markers.length > 0) {
            const { message, startLineNumber, startColumn } = markers[0];
            if (startLineNumber && startColumn) {
              errorMessage += `line[${startLineNumber}] column[${startColumn}]: `;
            }
            errorMessage += `Error. ${message}`;
          }
          if (errorMessage) {
            this.setState({ errorMessage: `Error. ${errorMessage}` });
            this.onChange(_editor.getValue(), false);
          }
          editor.setModelMarkers(model, 'default', markers ? markers : []);
        });
      });
    });
  }
}

const mapStateToProps = (state: AppState) => ({
  branding: selectBranding(state),
  devfileSchema: selectDevfileSchema(state),
  plugins: selectPlugins(state),
});

const connector = connect(mapStateToProps, null, null, { forwardRef: true });

type MappedProps = ConnectedProps<typeof connector>;
export default connector(DevfileEditor);
