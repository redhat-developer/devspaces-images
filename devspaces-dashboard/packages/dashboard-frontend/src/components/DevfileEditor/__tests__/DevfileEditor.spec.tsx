/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { Provider } from 'react-redux';
import DevfileEditor from '..';
import { BrandingData } from '../../../services/bootstrap/branding.constant';
import { FakeStoreBuilder } from '../../../store/__mocks__/storeBuilder';
import { createFakeCheWorkspace } from '../../../store/__mocks__/workspace';

jest.mock('monaco-editor-core', () => {
  return {
    LanguageConfiguration: typeof {},
    IMonarchLanguage: typeof {},
    Position: typeof {},
    IRange: typeof {},
    languages: {
      registerCompletionItemProvider: jest.fn(),
      registerDocumentSymbolProvider: jest.fn(),
      registerHoverProvider: jest.fn(),
      register: jest.fn(),
      setMonarchTokensProvider: jest.fn(),
      setLanguageConfiguration: jest.fn(),
      CompletionItemInsertTextRule: {
        InsertAsSnippet: jest.fn(),
      },
    },
    editor: {
      IModelDecoration: typeof [],
      create: jest.fn(),
      getModel: jest.fn(),
      getValue: jest.fn(),
      setModelMarkers: jest.fn(),
      defineTheme: jest.fn(),
      setTheme: jest.fn(),
    },
  };
});

describe('The DevfileEditor component', () => {
  it('should initialize the component correctly', () => {
    const workspaceName = 'wksp-test';
    const workspaceId = 'testWorkspaceId';
    const location = 'location[ \t]*(.*)[ \t]*$';
    const onChange = jest.fn();

    const component = renderComponent(workspaceName, workspaceId, location, onChange);

    expect(onChange).not.toBeCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });
});

function renderComponent(
  workspaceName: string,
  workspaceId: string,
  decorationPattern: string,
  onChange: (newValue: string, isValid: boolean) => void,
): ReactTestRenderer {
  const workspace = createFakeCheWorkspace(workspaceId, workspaceName);
  const store = new FakeStoreBuilder()
    .withCheWorkspaces({
      workspaces: [workspace],
    })
    .withBranding({
      docs: {
        devfile: 'devfile/documentation/link',
      },
    } as BrandingData)
    .build();
  return renderer.create(
    <Provider store={store}>
      <DevfileEditor
        devfile={workspace.devfile as che.WorkspaceDevfile}
        decorationPattern={decorationPattern}
        onChange={onChange}
      />
    </Provider>,
  );
}
