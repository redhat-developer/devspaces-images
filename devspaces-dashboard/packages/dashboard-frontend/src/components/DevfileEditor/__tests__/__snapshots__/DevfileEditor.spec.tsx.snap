// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The DevfileEditor component should initialize the component correctly 1`] = `
<div
  className="devfile-editor"
>
  <div
    className="monaco"
  >
    Â 
  </div>
  <div
    className="error"
  >
    
  </div>
  <a
    href="devfile/documentation/link"
    rel="noopener noreferrer"
    target="_blank"
  >
    Devfile Documentation
  </a>
</div>
`;
